bcda/database/connection.go

```
@@ -17,37 +17,50 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-var (
-	Connection *sql.DB
-	Pgxv5Pool  *pgxv5Pool.Pool
-)
-
-func init() {
+func Connect() *sql.DB {
 	cfg, err := LoadConfig()
-
 	if err != nil {
 		logrus.Fatalf("Failed to load database config %s", err.Error())
 	}
 
-	Connection, err = createDB(cfg)
+	conn, err := createDB(cfg)
 	if err != nil {
 		logrus.Fatalf("Failed to create db %s", err.Error())
 	}
 
-	Pgxv5Pool, err = CreatePgxv5DB(cfg)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	startConnectionHealthCheck(
+		ctx,
+		conn,
+		time.Duration(cfg.HealthCheckSec)*time.Second,
+	)
+
+	return conn
+}
```

We arent returning an error with this Connect() function, is that because any issues force a panic?
Can `conn` ever be a nil sql.DB object?  Like it returns properly but the actual connection isnt present (ie is more or less failing silently)?

---

This is a great question.
The current state: we don't return an error because issues force a process exit. This is how the previous `init()` worked and I didn't intend to change that.
The ideal state: _do_ we want to allow the connection to be nil? Is having a nil database ever a valid state? I would say no, since we are calling this function to connect to a database and I don't recall any instances in which we would ever want the result to be anything but a valid connection and not immediately panic. But this is super interesting and I'm open to exploring that.

---

Yeah, it looks like the health check does just log warnings, but in the event of an error in the `Connect()` and `ConnectPool() functions the process is exited via `logrus.Fatalf()`

---

I think that makes sense. If we are interested in exploring alternatives, I would also vote to do it in a separate PR/effort since I tried not to change too much of the _behavior_ of bcda in this PR.